[GavinSin] iP

GavinSin opened the pull request:
# Mira
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Mira frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

 - [X] Managing tasks
 - [ ] Managing deadlines (coming soon)
 - [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

GavinSin committed:
Refactor class Duke to Mira.

This is to reflect the new chat bot name with new welcome messages.

Other files with word 'Duke' is replaced with 'Mira' as well.

GavinSin committed:
Create new class UI (User Interface)

- readCommand, echoCommand, showMessage are related methods dealing
with user input and output

- the main logics of chatbot code remains on Mira class

GavinSin committed:
Create TaskList class

This handles addTask and listTask methods

GavinSin committed:
Reuse UI class in TaskList class

Using UI class and methods for all user interactions, it provides
centralized control of the output styling such as printLine

GavinSin committed:
Create new Task class

It is more natural to represent task in TaskList as class instead
of primitive values.

GavinSin committed:
Update TaskList to use Task class

markTask and unmarkTask methods are added in TaskList class

GavinSin committed:
Update Mira class with mark and unmark commands

GavinSin committed:
Create subclass Deadline, Event, and Todo

This is so that mentioned class contains shared Task methods, and
having differing behaviors via inheritance and polymorphism

GavinSin committed:
Update Mira class with new commands

update addTask method with input Task to take advantage of
polymorphism

GavinSin committed:
Update input and output file for text UI testing

GavinSin committed:
Replaced incorrect inputs message via Exceptions

GavinSin committed:
Rewrite TaskList class with ArrayList implementation and add delete
method

GavinSin committed:
Make all subclasses of Task have toFileString method

toFileString method returns a formatted string to be used for saving
of tasks

GavinSin committed:
Add Storage class

Given a relative file path, the Storage class can save and load
tasks in specified string format

GavinSin committed:
Update Mira and TaskList class to save tasks in disk automatically
whenever the task list changes

saveTask method used in creation of new task where content is
written in append mode

saveTasks method used for update and delete of existing task where
content is overwritten

GavinSin committed:
Merge branch 'branch-Level-7'

GavinSin committed:
Use LocalDateTime for datetime of Event and Deadline tasks

GavinSin committed:
Merge branch 'branch-Level-8'

GavinSin committed:
Rename class UI to Ui

GavinSin committed:
Add Parser, ExitCommand, ListCommand classes

GavinSin committed:
Add MarkCommand, UnmarkCommand, DeleteCommand and TodoCommand class

GavinSin committed:
Add EventCommand and DeadlineCommand

GavinSin committed:
Refactor Mira constructor to handle exception internally

GavinSin committed:
Add JavaDoc comments to all Command classes and Savable interface

GavinSin committed:
Merge branch 'branch-A-MoreOOP'

GavinSin committed:
Move Command and its subclasses into mira.command package, the rest
is in mira package

GavinSin committed:
Merge branch 'add-gradle-support'

GavinSin committed:
Setup gradle and run Mira with it

GavinSin committed:
Add test cases: ParserTest, StorageTest, MarkCommandTest

GavinSin committed:
Fix test data for JUnit testing

GavinSin committed:
Reformat JavaDoc comments according to specified standard

GavinSin committed:
Merge branch 'branch-A-JavaDoc'

GavinSin committed:
Reformat code to match checkstyle.xml standard

GavinSin committed:
Merge branch 'branch-A-CodingStandard'

GavinSin committed:
Add missing JavaDoc comments in Delete and Event Command

GavinSin committed:
Add find keyword feature

GavinSin committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/mira/Parser.java

GavinSin committed:
Fix additional JavaDoc comments after Level-9 merge commit

S-K-Y-Light commented on src/main/java/mira/Event.java:
```diff
@@ -0,0 +1,73 @@
+
+    /**
+     * Constructs a new Event task.
+     * It contains specified description, start, and end datetime.
```
Small Nitpick: Maybe "start" could be renamed to "start datetime" for greater clarity?
Same Issue for the other Documentation in Event.java

S-K-Y-Light commented on src/main/java/mira/Mira.java:
```diff
@@ -0,0 +1,87 @@
+     * </p>
+     *
+     * @param filePath The path to the file where tasks will be stored and loaded from.
+     *                 If the file does not exist, it will be created.
```
I like that you did HTML Formatting in your Documentation 👍

S-K-Y-Light commented on src/main/java/mira/Parser.java:
```diff
@@ -0,0 +1,75 @@
+ */
+public class Parser {
+    /**
+     * Parses the user's input command and returns the corresponding {@link Command} object.
+     * <p>
+     * The method splits the input string into the command and its arguments,
+     * and then maps it to the appropriate command class, such as {@link MarkCommand},
+     * {@link UnmarkCommand}, {@link DeadlineCommand}, etc.
+     * </p>
+     *
+     * @param fullCommand The full command string entered by the user.
+     * @return The {@link Command} object corresponding to the user's input.
+     * @throws MiraException If the command is not recognized or if the command arguments
+     *         are in an invalid format.
```
I like that you're using @link for JavaDocs 👍

S-K-Y-Light commented on src/main/java/mira/Storage.java:
```diff
@@ -0,0 +1,102 @@
+     * @throws IOException If an I/O error occurs during loading.
+     * @throws SecurityException If a permission error occurs during loading.
+     * @throws IllegalArgumentException If file content corrupts during loading.
+     * @throws MiraException If unsupported task type founds during loading.
```
Small Nitpick: Is there a Grammatical Error? I think "founds" should be "found".

S-K-Y-Light commented on src/main/java/mira/Task.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    /**
+     * Return the string format of task to be saved for storage.
```
Maybe "Return" should be "Returns"? Just to keep consistency with the other documentations.

S-K-Y-Light commented on src/main/java/mira/TaskList.java:
```diff
@@ -0,0 +1,141 @@
+ * TaskList handles adding and listing of tasks.
+ */
+public class TaskList {
+    /* Array to store tasks */
```
Small Nitpick: I think "/*" should start with "/**".

S-K-Y-Light commented on src/test/java/mira/StorageTest.java:
I think your other test files start documentation with "Tests" instead of "Test". Maybe pick one to keep consistency?

S-K-Y-Light commented on src/main/java/mira/Task.java:
I noticed that for your documentation, you added indentations for @param information for some files but not this one.
Same Issue with some other files. Maybe keep to one style (with indentations or without indentations)?

S-K-Y-Light reviewed (commented):
Mostly Documentation Issues. Otherwise, I really like your code. LGTM.

GavinSin committed:
Fix check style import issue

GavinSin committed:
Merge branch 'branch-A-CheckStyle'

GavinSin committed:
Replace text UI with GUI

GavinSin committed:
Add a placeholder for TextField

GavinSin committed:
Set title on the application

GavinSin committed:
Delete unused text Ui class

GavinSin committed:
Merge branch 'branch-Level-10'

weibinlim commented on src/test/java/mira/ParserTest.java:
```diff
@@ -0,0 +1,57 @@
+     * Tests that `parse` throws `MiraException` when an unknown command is provided.
+     */
+    @Test
+    public void testParse_unknownCommand_throwsMiraException() {
```
I like that your method names are very meaningful!

weibinlim commented on src/main/java/Mira.java:
```diff
@@ -3,17 +3,28 @@
-    private boolean isRunning; // default is true
+    private final Ui ui; // handle user interface
+    private final Storage storage; // storage for saving, loading tasks
+    private TaskList tasks; // Manages tasks
```
I like that you used the plural of task to name the list of tasks!

weibinlim commented on src/main/java/TaskList.java:
```diff
@@ -47,15 +47,15 @@ public void deleteTask(int index) throws IOException {
      * Lists all the tasks currently stored in the task list.
      */
-    public void listTasks() {
+    public String listTasks() {
```
I like how the method name describes its function!

weibinlim commented on src/main/java/mira/Storage.java:
```diff
@@ -89,14 +89,14 @@ private Task createTaskFromFile(String[] parts) throws MiraException {
         String description = parts[2];
```
Perhaps you could use a more descriptive name for the split description of the task?

weibinlim reviewed (commented):
Really liked your code as it was very comprehensive and everything was easy to read and understand!

GavinSin committed:
Add assertions on Mira class

Mira class `getResponse` method is an important **happy** path for
the application to work correctly.

Assumptions that are tested:
* user input is not null
* `TaskList` is initialized

GavinSin committed:
Refactor section of codes via the use of guard clauses

By using guard clauses, the happy paths in the methods `loadTasks`
and `findTasks` are emphasized, making it more readable.

Appropriate empty lines are used to organize sections of code.

GavinSin committed:
Merge pull request #1 from GavinSin/branch-A-Assertions

Add assertions on Mira class

GavinSin committed:
Merge branch 'master' of https://github.com/GavinSin/ip

* 'master' of https://github.com/GavinSin/ip:
  Add assertions on Mira class

GavinSin committed:
Merge branch 'master' into branch-A-CodeQuality

GavinSin committed:
Merge pull request #2 from GavinSin/branch-A-CodeQuality

Branch a code quality

GavinSin committed:
Add expense feature to existing app

Similar to TaskList and Todo task class, Expense and ExpenseList
are created to follow similar structure, and existing logics can
be reused.

GavinSin committed:
Merge pull request #3 from GavinSin/branch-D-Expenses

Add expense feature to existing app

GavinSin committed:
Add user guide for Mira chatbot

GavinSin committed:
Merge branch 'branch-A-UserGuide'

GavinSin committed:
Fix header comment for Task method

GavinSin committed:
Fix bye command action to exit application
